pkg/orchestrator/housekeeping.go

```diff
@@ -0,0 +1,219 @@
+		validate.IsGreaterThanZero(params.Workers, "workers must be greater than zero"),
+		validate.IsGreaterThanZero(params.TimeoutBuffer, "timeout buffer must be greater than zero"),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error validating housekeeping params. %w", err)
+	}
+
+	h := &Housekeeping{
+		evaluationBroker: params.EvaluationBroker,
+		jobStore:         params.JobStore,
+		interval:         params.Interval,
+		timeoutBuffer:    params.TimeoutBuffer,
+		workersSem:       make(chan struct{}, params.Workers),
+	}
+
+	return h, nil
+}
+
+// ShouldRun returns true if the housekeeping task should run.
+// This is just a placeholder for now until we introduce leader election or lease management for housekeeping
+// when we introduce more than one orchestrator.
+func (h *Housekeeping) ShouldRun() bool {
+	return true
+}
+
+// Start starts the housekeeping task
+func (h *Housekeeping) Start(ctx context.Context) {
+	h.startOnce.Do(func() {
+		h.ctx, h.cancel = context.WithCancel(ctx)
+		go h.runHousekeepingTasks()
+	})
+}
+
+func (h *Housekeeping) Stop(ctx context.Context) {
+	if h.cancel != nil {
+		h.cancel()
+	}
+
+	// wait for inflight housekeeping tasks to complete, or until the context is done
+	waitGroupDone := make(chan struct{})
+	go func() {
+		h.waitGroup.Wait()
+		close(waitGroupDone)
+	}()
+
+	select {
+	case <-waitGroupDone:
+	case <-ctx.Done():
+		h.waitGroup.Done()
+	}
+}
+
+func (h *Housekeeping) runHousekeepingTasks() {
+	ticker := time.NewTicker(h.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if !h.ShouldRun() {
+				continue
+			}
+
+			// fetch active executions
+			activeExecutions := h.fetchActiveExecutions()
+
+			// run housekeeping tasks
+			h.timeoutExecutions(activeExecutions)
+		case <-h.ctx.Done():
+			log.Ctx(h.ctx).Debug().Msg("stopped housekeeping task")
+			return
+		}
+	}
+}
+
+// fetchActiveExecutions fetches all active executions
+func (h *Housekeeping) fetchActiveExecutions() []*models.Execution {
+	var activeExecutions []*models.Execution
+	activeJobs, err := h.jobStore.GetInProgressJobs(h.ctx, "")
+	if err != nil {
+		log.Ctx(h.ctx).Err(err).Msg("failed to get active jobs")
+		return activeExecutions
+	}
+
+	for i := range activeJobs {
+		job := &activeJobs[i]
+
+		// we only have housekeeping tasks for batch and ops jobs
+		if job.IsLongRunning() {
+			continue
+		}
+		executions, err := h.jobStore.GetExecutions(h.ctx, jobstore.GetExecutionsOptions{
+			JobID: job.ID,
+		})
+		if err != nil {
+			// log error and avoid having a single job failure affect the housekeeping of other jobs
+			log.Ctx(h.ctx).Err(err).Msgf("failed to get executions for job %s", job.ID)
+			continue
+		}
+		// filter terminal executions, and enrich executions with job information
+		for j := range executions {
+			if executions[j].IsTerminalState() {
+				continue
+			}
+			executions[j].Job = job
+			activeExecutions = append(activeExecutions, &executions[j])
+		}
+	}
```

Comment:
While I don't think there is anything wrong here, I am dubious of taking the address of variables in loops, especially nested loops. See https://go.dev/blog/loopvar-preview for details.

Blocking:
 Let's make sure there is test coverage for cases when there are multiple jobs with multiple executions - at present this scenario has no coverage.

Reply:
I am happy they are fixing this soon. In my case, I am looping over indexes and fetching the reference from the original array to avoid failing into the issue of referencing the wrong pointer

Added test in https://github.com/bacalhau-project/bacalhau/pull/3974/commits/4fff4d01760ea9be8c1a03e8ef318a2b792af16b